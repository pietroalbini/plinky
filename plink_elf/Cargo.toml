[package]
name = "plink_elf"
version = "0.1.0"
edition = "2021"

[dependencies]
plink_macros = { path = "../plink_macros" }

[dev-dependencies]
anyhow = "1.0.71"
insta = "1.30.0"
tempfile = "3.6.0"
