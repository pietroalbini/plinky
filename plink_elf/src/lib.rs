pub mod errors;
mod reader;
mod types;
mod utils;
mod writer;

pub use self::types::*;
