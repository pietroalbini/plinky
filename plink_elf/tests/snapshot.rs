//! Check whether ELF files are parsed correctly.

use anyhow::Error;
use plink_elf::Object;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use std::process::Command;
use tempfile::NamedTempFile;

macro_rules! test {
    ($name:ident, $source:expr $(, $variant:ident)*) => {
        mod $name {
            use super::*;

            $(
                #[test]
                fn $variant() -> Result<(), Error> {
                    implement_test($source, stringify!($variant))
                }
            )*
        }
    };
}

test!(
    hello_asm,
    "tests/snapshot/hello_asm.S",
    x86,
    x86_64,
    x86_linked,
    x86_64_linked
);
test!(hello_c, "tests/snapshot/hello_c.c", x86, x86_64);

#[track_caller]
fn implement_test(source: &str, variant: &str) -> Result<(), Error> {
    let (variant, link) = match variant {
        "x86" => (Variant::X86, Link::No),
        "x86_64" => (Variant::X86_64, Link::No),
        "x86_linked" => (Variant::X86, Link::Yes),
        "x86_64_linked" => (Variant::X86_64, Link::Yes),
        other => panic!("unsupported variant: {other}"),
    };

    let mut object_file = match source.rsplit_once('.').map(|(_name, ext)| ext) {
        Some("S") => compile_asm(source, variant)?,
        Some("c") => compile_c(source, variant)?,
        Some(other) => panic!("unsupported extension: {other}"),
        None => panic!("missing extension for {source}"),
    };
    if let Link::Yes = link {
        object_file = link_single_object(object_file.path(), variant)?;
    }
    let mut file = BufReader::new(File::open(object_file.path())?);

    let parsed = Object::load(&mut file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_omit_expression(true);
    settings.set_snapshot_path("snapshot");
    settings.set_prepend_module_to_snapshot(false);
    settings.set_snapshot_suffix(match (variant, link) {
        (Variant::X86, Link::No) => "32bit",
        (Variant::X86_64, Link::No) => "64bit",
        (Variant::X86, Link::Yes) => "32bit-linked",
        (Variant::X86_64, Link::Yes) => "64bit-linked",
    });
    let _guard = settings.bind_to_scope();

    let name = source
        .rsplit_once('/')
        .map(|(_dir, name)| name)
        .unwrap_or(source);
    let name = name
        .rsplit_once('.')
        .map(|(name, _ext)| name)
        .unwrap_or(name);

    insta::assert_snapshot!(name, format!("{parsed:#x?}"));
    Ok(())
}

fn compile_asm(source: &str, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("nasm")
        .arg("-o")
        .arg(dest.path())
        .arg("-f")
        .arg(match variant {
            Variant::X86 => "elf32",
            Variant::X86_64 => "elf64",
        })
        .arg(source)
        .status()?;
    anyhow::ensure!(status.success(), "failed to compile {}", source);
    Ok(dest)
}

fn compile_c(source: &str, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("gcc")
        .arg("-c")
        .arg("-o")
        .arg(dest.path())
        .arg(source)
        .arg(match variant {
            Variant::X86 => "-m32",
            Variant::X86_64 => "-m64",
        })
        .status()?;
    anyhow::ensure!(status.success(), "failed to compile {}", source);
    Ok(dest)
}

fn link_single_object(object: &Path, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("ld")
        .arg("-o")
        .arg(dest.path())
        .arg(object)
        .args(match variant {
            Variant::X86 => ["-m", "elf_i386"],
            Variant::X86_64 => ["-m", "elf_x86_64"],
        })
        .status()?;
    anyhow::ensure!(status.success(), "failed to link {object:?}");
    Ok(dest)
}

#[derive(Clone, Copy)]
enum Variant {
    X86,
    X86_64,
}

#[derive(Clone, Copy)]
enum Link {
    Yes,
    No,
}
