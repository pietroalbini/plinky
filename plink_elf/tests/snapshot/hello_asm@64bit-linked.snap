---
source: plink_elf/tests/snapshot.rs
---
ElfObject {
    env: ElfEnvironment {
        class: Elf64,
        endian: Little,
        abi: SystemV,
        machine: X86_64,
    },
    type_: Executable,
    entry: Some(
        0x401000,
    ),
    flags: 0x0,
    sections: {
        "#0 ": ElfSection {
            name: "",
            memory_address: 0x0,
            content: Null,
        },
        "#1 .text": ElfSection {
            name: ".text",
            memory_address: 0x401000,
            content: Program(
                ElfProgramSection {
                    writeable: false,
                    allocated: true,
                    executable: true,
                    raw: RawBytes {
                        b8 04 00 00 00 bb 01 00 00 00 b9 00 20 40 00 ba 0d 00 00 00 cd 80 b8 04 00 | ............ @...........
                        00 00 bb 01 00 00 00 b9 0d 20 40 00 ba 0f 00 00 00 cd 80 b0 01 bb 00 00 00 | ......... @..............
                        00 cd 80                                                                   | ...
                        (len: 0x35)
                    },
                },
            ),
        },
        "#2 .data": ElfSection {
            name: ".data",
            memory_address: 0x402000,
            content: Program(
                ElfProgramSection {
                    writeable: true,
                    allocated: true,
                    executable: false,
                    raw: RawBytes {
                        48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a 47 6f 6f 64 62 79 65 20 77 6f 72 6c | Hello world!.Goodbye worl
                        64 21 0a                                                                   | d!.
                        (len: 0x1c)
                    },
                },
            ),
        },
        "#3 .symtab": ElfSection {
            name: ".symtab",
            memory_address: 0x0,
            content: SymbolTable(
                ElfSymbolTable {
                    symbols: {
                        "#00 <empty>": ElfSymbol {
                            name: "",
                            binding: Local,
                            type_: NoType,
                            definition: Undefined,
                            value: 0x0,
                            size: 0x0,
                        },
                        "#01 tests/snapshot/hello_asm.S": ElfSymbol {
                            name: "tests/snapshot/hello_asm.S",
                            binding: Local,
                            type_: File,
                            definition: Absolute,
                            value: 0x0,
                            size: 0x0,
                        },
                        "#02 msg_hello": ElfSymbol {
                            name: "msg_hello",
                            binding: Local,
                            type_: NoType,
                            definition: Section(
                                "#2 .data",
                            ),
                            value: 0x402000,
                            size: 0x0,
                        },
                        "#03 len_hello": ElfSymbol {
                            name: "len_hello",
                            binding: Local,
                            type_: NoType,
                            definition: Absolute,
                            value: 0xd,
                            size: 0x0,
                        },
                        "#04 msg_goodbye": ElfSymbol {
                            name: "msg_goodbye",
                            binding: Local,
                            type_: NoType,
                            definition: Section(
                                "#2 .data",
                            ),
                            value: 0x40200d,
                            size: 0x0,
                        },
                        "#05 len_goodbye": ElfSymbol {
                            name: "len_goodbye",
                            binding: Local,
                            type_: NoType,
                            definition: Absolute,
                            value: 0xf,
                            size: 0x0,
                        },
                        "#06 _start": ElfSymbol {
                            name: "_start",
                            binding: Global,
                            type_: NoType,
                            definition: Section(
                                "#1 .text",
                            ),
                            value: 0x401000,
                            size: 0x0,
                        },
                        "#07 __bss_start": ElfSymbol {
                            name: "__bss_start",
                            binding: Global,
                            type_: NoType,
                            definition: Section(
                                "#2 .data",
                            ),
                            value: 0x40201c,
                            size: 0x0,
                        },
                        "#08 _edata": ElfSymbol {
                            name: "_edata",
                            binding: Global,
                            type_: NoType,
                            definition: Section(
                                "#2 .data",
                            ),
                            value: 0x40201c,
                            size: 0x0,
                        },
                        "#09 _end": ElfSymbol {
                            name: "_end",
                            binding: Global,
                            type_: NoType,
                            definition: Section(
                                "#2 .data",
                            ),
                            value: 0x402020,
                            size: 0x0,
                        },
                    },
                },
            ),
        },
        "#4 .strtab": ElfSection {
            name: ".strtab",
            memory_address: 0x0,
            content: StringTable(
                ElfStringTable {
                    strings: {
                        0x0: "",
                        0x1: "tests/snapshot/hello_asm.S",
                        0x1c: "msg_hello",
                        0x26: "len_hello",
                        0x30: "msg_goodbye",
                        0x3c: "len_goodbye",
                        0x48: "__bss_start",
                        0x54: "_edata",
                        0x5b: "_end",
                    },
                },
            ),
        },
        "#5 .shstrtab": ElfSection {
            name: ".shstrtab",
            memory_address: 0x0,
            content: StringTable(
                ElfStringTable {
                    strings: {
                        0x0: "",
                        0x1: ".symtab",
                        0x9: ".strtab",
                        0x11: ".shstrtab",
                        0x1b: ".text",
                        0x21: ".data",
                    },
                },
            ),
        },
    },
    segments: [
        ElfSegment {
            content: Unknown {
                id: 0x1,
                raw: RawBytes {
                    7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 02 00 3e 00 01 00 00 00 00 | .ELF..............>......
                    10 40 00 00 00 00 00 40 00 00 00 00 00 00 00 98 21 00 00 00 00 00 00 00 00 | .@.....@........!........
                    00 00 40 00 38 00 03 00 40 00 06 00 05 00 01 00 00 00 04 00 00 00 00 00 00 | ..@.8...@................
                    00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 40 00 00 00 00 00 e8 00 00 00 | .......@.......@.........
                    00 00 00 00 e8 00 00 00 00 00 00 00 00 10 00 00 00 00 00 00 01 00 00 00 05 | .........................
                    00 00 00 00 10 00 00 00 00 00 00 00 10 40 00 00 00 00 00 00 10 40 00 00 00 | .............@.......@...
                    00 00 35 00 00 00 00 00 00 00 35 00 00 00 00 00 00 00 00 10 00 00 00 00 00 | ..5.......5..............
                    00 01 00 00 00 06 00 00 00 00 20 00 00 00 00 00 00 00 20 40 00 00 00 00 00 | .......... ....... @.....
                    00 20 40 00 00 00 00 00 1c 00 00 00 00 00 00 00 1c 00 00 00 00 00 00 00 00 | . @......................
                    10 00 00 00 00 00 00                                                       | .......
                    (len: 0xe8)
                },
            },
        },
        ElfSegment {
            content: Unknown {
                id: 0x1,
                raw: RawBytes {
                    b8 04 00 00 00 bb 01 00 00 00 b9 00 20 40 00 ba 0d 00 00 00 cd 80 b8 04 00 | ............ @...........
                    00 00 bb 01 00 00 00 b9 0d 20 40 00 ba 0f 00 00 00 cd 80 b0 01 bb 00 00 00 | ......... @..............
                    00 cd 80                                                                   | ...
                    (len: 0x35)
                },
            },
        },
        ElfSegment {
            content: Unknown {
                id: 0x1,
                raw: RawBytes {
                    48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a 47 6f 6f 64 62 79 65 20 77 6f 72 6c | Hello world!.Goodbye worl
                    64 21 0a                                                                   | d!.
                    (len: 0x1c)
                },
            },
        },
    ],
}
